File: 329-longest-increasing-path-in-a-matrix.java
import java.util.Arrays;

class Solution {
    public int longestIncreasingPath(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        int max = 0;
        int[][] record = new int[m][n];

        for (int[] row : record) {
            Arrays.fill(row, -1);
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                max = Math.max(max, dfs(i, j, 0, matrix, record));
            }
        }
        return max;
    }

    private int dfs(int i, int j, int lastNum, int[][] matrix, int[][] record) {
        if (i < 0 || j < 0 || i >= matrix.length || j >= matrix[0].length || matrix[i][j] <= lastNum) {
            return 0;
        }

        if (record[i][j] > -1) {
            return record[i][j];
        }

        int up = dfs(i - 1, j, matrix[i][j], matrix, record);
        int down = dfs(i + 1, j, matrix[i][j], matrix, record);
        int left = dfs(i, j - 1, matrix[i][j], matrix, record);
        int right = dfs(i, j + 1, matrix[i][j], matrix, record);

        int pathNum = 1 + max(up, down, left, right);
        record[i][j] = pathNum;
        return pathNum;
    }

    private int max(int a, int b, int c, int d) {
        return Math.max(Math.max(a, b), Math.max(c, d));
    }
}

File: python/329-longest-increasing-path-in-a-matrix.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        m = len(matrix)
        n = len(matrix[0])
        cache = [[-1 for j in range(m)] for i in range(n)]
        directions = [[1,0],[-1,0],[0,-1],[0,1]]

        def dfs(i, j, target) -> int:
            if i==-1 or j==-1 or i==m or j==n or matrix[i][j]<=target:
                return 0
            
            if cache[i][j]>-1:
                return cache[i][j]
            
            for direction in directions:
                cache[i][j] = 1 + max(dfs(i-1, j, matrix[i][j]), dfs(i+1,j, matrix[i][j]),dfs(i, j-1, matrix[i][j]), dfs(i, j+1, matrix[i][j]))
            return  cache[i][j]
            
        ans = 1;
        for i in range(m):
            for j in range(n):
                ans= max(dfs(i, j, -1), ans)
                
        return ans
    